### PR TITLE
ARROW-18436: [C++] Ensure correct (un)escaping of special characters in URI paths

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@westonpace @vibhatha This may affect Substrait, though it should be for the better.

**Comment by pitrou** [comment]
@jorisvandenbossche @AlenkaF This sanitizes the `filesystem_from_uri` behavior (see unit tests).

**Review by lidavidm — APPROVED** [review]

**Comment by vibhatha** [comment]
> @westonpace @vibhatha This may affect Substrait, though it should be for the better.

looks good to me, @westonpace do you see any issues?

File: cpp/src/arrow/util/uri.cc [hunk 1]
```diff
@@ -58,10 +59,10 @@ bool IsDriveSpec(const std::string_view s) {
 
 }  // namespace
 
-std::string UriEscape(const std::string& s) {
+std::string UriEscape(const std::string_view& s) {
```
Comment (westonpace) [inline cpp/src/arrow/util/uri.cc:62]: Minor nit: I think we've been preferring to pass `string_view` by value in Substrait per https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/ and @bkietz 's advice

File: cpp/src/arrow/util/uri.cc [hunk 2]
```diff
@@ -306,14 +311,29 @@ Status Uri::Parse(const std::string& uri_string) {
...
+  }
+  std::string out;
 #ifdef _WIN32
-  // Path is supposed to start with "X:/..."
-  return "file:///" + path;
+  // Turn "/" separators into "\", as Windows recognizes both but uriparser
+  // only the latter.
+  std::string fixed_path(path);
+  std::replace(fixed_path.begin(), fixed_path.end(), '/', '\\');
+  out.resize(8 + 3 * fixed_path.length() + 1);
```
Comment (westonpace) [inline cpp/src/arrow/util/uri.cc:325]: Minor nit: Maybe a comment explaining this math?  Probably not needed if this is a normal Windows thing.

**Review by westonpace — COMMENTED** [review]
I agree it's for the better, thanks.  A few small thoughts.

File: cpp/src/arrow/util/uri.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/uri.cc:62]: Ah, good point, thanks.

File: cpp/src/arrow/util/uri.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/uri.cc:325]: This is simply as per the `uriWindowsFilenameToUriStringA` doc:
https://uriparser.github.io/doc/api/latest/Uri_8h.html#a422dc4a2b979ad380a4dfe007e3de845

File: cpp/src/arrow/util/uri.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/uri.cc]: I opened https://github.com/apache/arrow/issues/15171 for other instances of this.

**Pull request merged**